From cbb9b1ba6eb981ff5276f37cc35d52ba9a7d4c08 Mon Sep 17 00:00:00 2001
From: Ivan Efremov <i.efremov@yadro.com>
Date: Thu, 14 May 2020 23:45:06 +0300
Subject: [PATCH 57/57] scsi: qla2xxx: Reduce timeout if cable is unplugged

Target creation (i.e. write of "1" to "tpg_1/enable") blocks for 23
seconds if cable is not connected. Firmware always returns "loss of
sync" state during the period. It slows down management software.

If there's a cable, firmware may not report "loss of sync" state at all.
E.g. in the initiator mode, the first state reported after driver
initialization may be "wait login". Then, within 500 ms, it reaches
"ready" state. In pure target mode it was observed that firmware goes
directly to the "ready" state if there's cable.

Therefore it should be safe to unblock configfs much earlier if firmware
stays in "loss of sync" only for 5 seconds. Total execution time of the
write to "tpg_1/enable" is ~4 seconds if there's a cable and ~9 seconds
otherwise.

Signed-off-by: Ivan Efremov <i.efremov@yadro.com>
Signed-off-by: Roman Bolshakov <r.bolshakov@yadro.com>
---
 drivers/scsi/qla2xxx/qla_init.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/scsi/qla2xxx/qla_init.c b/drivers/scsi/qla2xxx/qla_init.c
index 111111111..111111111 100644
--- a/drivers/scsi/qla2xxx/qla_init.c
+++ b/drivers/scsi/qla2xxx/qla_init.c
@@ -4540,7 +4540,7 @@ static int
 qla2x00_fw_ready(scsi_qla_host_t *vha)
 {
 	int		rval;
-	unsigned long	wtime, mtime, cs84xx_time;
+	unsigned long	wtime, mtime, cs84xx_time, ltime;
 	uint16_t	min_wait;	/* Minimum wait time if loop is down */
 	uint16_t	wait_time;	/* Wait time if loop is coming ready */
 	uint16_t	state[6];
@@ -4563,6 +4563,9 @@ qla2x00_fw_ready(scsi_qla_host_t *vha)
 		wait_time = min_wait;
 	}
 
+	/* timeout for loss of sync state */
+	ltime = jiffies + (5 * HZ);
+
 	/* Min wait time if loop down */
 	mtime = jiffies + (min_wait * HZ);
 
@@ -4626,7 +4629,9 @@ qla2x00_fw_ready(scsi_qla_host_t *vha)
 				/* Loop down. Timeout on min_wait for states
 				 * other than Wait for Login.
 				 */
-				if (time_after_eq(jiffies, mtime)) {
+				if ((state[0] == FSTATE_LOSS_OF_SYNC &&
+				    time_after_eq(jiffies, ltime)) ||
+				    time_after_eq(jiffies, mtime)) {
 					ql_log(ql_log_info, vha, 0x8038,
 					    "Cable is unplugged...\n");
 
